src/webhook_launcher/app/models.py

```diff
@@ -392,24 +399,119 @@ def clean(self, exclude=None):
                     (project, self.user)
                 )
 
-    def trigger_build(self):
-        if not self.lsr:
-            mylsr, created = LastSeenRevision.objects.get_or_create(
-                mapping=self,
+    def trigger_build(self, user=None, tag=None, force=False):
+
+        # Only fire for projects which allow webhooks. We can't just
+        # rely on validation since a Project may forbid hooks after
+        # the hook was created
+        if self.project_disabled:
+            print "Project has build disabled"
+            return
+
+        handled = self.lsr.handled and self.lsr.tag == tag and not force
```

Comment:
In theory yes, but in practice trigger_build is always called on saved WHM object, so the lsr property will return a LSR object. But could add a check there that the LSR object is valid.